lua/ttt2/libraries/orm.lua

```diff
@@ -0,0 +1,267 @@
+	-- Most likely rowids wont change in gmod as there is no vacuum operation but just to be safe we will not allow to use such tables. ref: https://wiki.facepunch.com/gmod/sql
+
+	if not self._primaryKey then
+		return object
+	end
+
+	object.Delete = ORMOBJECT.Delete
+	object.Save = ORMOBJECT.Save
+	object.Refresh = ORMOBJECT.Refresh
+	object._primaryKey = self._primaryKey
+	object._primaryKeyList = self._primaryKeyList
+	object._columnList = self._columnList
+
+	return object
+end
+
+---
+-- Retrieves all saved objects of the model with the given filters from the database.
+-- @param table filters An array of filters. Each filter should contain a `column`, `op`, `value` and `concat`(if it is not the last filter).
+-- @return table|nil Returns an array of all found objects or nil in case of an error.
+-- @realm shared
+function ORMMODEL:Where(filters)
+	local query = "SELECT * FROM " .. sql.SQLIdent(self._tableName) .. " WHERE "
+	local whereList = {}
+
+	for i = 1, #filters do
+		local curFilter = filters[i]
+
+		whereList[i] = sql.SQLIdent(curFilter[column]) .. (curFilter[op] or "=") .. sql.SQLStr(curFilter[value]) .. (curFilter[concat] or "")
+	end
+
+	whereList = table.concat(whereList)
+
+	query = query .. whereList
+
+	local objects = sql.Query(query)
+
+	if objects == false then return end
+
+	if not self._primaryKey then
+		return objects
+	end
+
+	for i = 1, #objects do
+		objects[i].Save = ORMOBJECT.Save
+		objects[i].Delete = ORMOBJECT.Delete
+		objects[i].Refresh = ORMOBJECT.Refresh
+	end
+
+	return objects or {}
+end
+
+---
+-- @class ORMOBJECT
+
+---
+-- Deletes the given object from the database storage.
+-- @return boolean Returns true if the object was successfully deleted, false otherwise.
+-- @realm shared
+function ORMOBJECT:Delete()
+	local where = {}
+	local primaryKey = self._primaryKey
+
+	for i = 1, #primaryKey do
+		where[i] = sql.SQLIdent(primaryKey[i]) .. "=" .. sql.SQLStr(self[primaryKey[i]])
+	end
+
+	where = table.concat(where, " AND ")
+
+	local query = "DELETE FROM " .. sql.SQLIdent(self._tableName) .. " WHERE " .. where
+
+	return sql.Query(query) == nil
+end
+
+---
+-- Saves the data of the given object to the database storage.
+-- @return boolean Returns true if the object was successfully saved to the database, false otherwise.
+-- @realm shared
+function ORMOBJECT:Save()
+	local query = "INSERT INTO " .. sql.SQLIdent(self._tableName) .. "("
+	local valueList = {nil, nil}
+	local dataStructure = self._dataStructure
+
+	for i = 1, #dataStructure do
+		valueList[i] = sql.SQLStr(self[dataStructure[i]])
+	end
+
+	valueList = table.concat(valueList, ",")
+
+	query = query .. self._columnList .. ") VALUES(" .. valueList .. ") ON CONFLICT(" .. self._primaryKeyList .. ") DO UPDATE SET(" .. self._columnList .. ")=(" .. valueList .. ");"
+
+	return sql.Query(query) == nil
```

Comment:
I know, it looks weird, but I'd prefer the following to avoid issues if sql returns changes and returns an empty table too for an empty result:

```suggestion
	return not sql.Query(query) == false
```